### PR TITLE
implement tacacs_global to match radius_global

### PR DESCRIPTION
**Pull request by tphoney**

### CODE REVIEW

**Comment by tphoney** [comment]
Node under test:
- name  - ldv6kkd8odvbueu
- type  - R-IOSXRV9000-CH
- image - 6.0.0.20I

TestTacacsGlobal#test_tacacs_global = 31.99 s = .

Finished in 37.430808s, 0.0267 runs/s, 0.2939 assertions/s.

**Comment by tphoney** [comment]
@chburket @mikewiebe i have added a provided that more closely mimics radius_global. This makes work on the puppet module easier.

**Comment by tphoney** [comment]
# Running:

Node under test:
- name  - switch
- type  - N9K-NXOSV
- image - bootflash:///nxos.7.0.3.I2.1.bin

TestTacacsGlobal#test_tacacs_global = 12.81 s = .

Finished in 14.139052s, 0.0707 runs/s, 0.7780 assertions/s.

File: lib/cisco_node_utils/cmd_ref/tacacs_global.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# enable_global.yaml
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/tacacs_global.yaml]: `/enable/tacacs/`

File: lib/cisco_node_utils/tacacs_global.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# Tacacs Global provider class
+
+# TP HONEY et al., June 2016
+
+# Copyright (c) 2014-2016 Cisco and/or its affiliates.
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/tacacs_global.rb:5]: `/2014-2016/2016/` 

Fix other files too.

File: lib/cisco_node_utils/cmd_ref/tacacs_global.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  default_value: ""
+  nexus:
+    get_command: "show run tacacs all"
+  ios_xr:
+    get_command: "show running-config tacacs-server"
+
+key_format:
+  auto_default: false
+  get_value: '/^tacacs-server key (\d+)\s+(\S+)/'
+  default_value: 0
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/tacacs_global.yaml:16]: You're using both `auto_default` and `default_value` ? This should be a one-or-the-other.

File: lib/cisco_node_utils/tacacs_global.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        config_set('tacacs_global',
+                   'timeout',
+                   state:   '',
+                   timeout: val)
+      end
+    end
+
+    def key_format
+      match = config_get('tacacs_global', 'key_format')
+      match.nil? ? TACACS_GLOBAL_ENC_UNKNOWN : match[0].to_i
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/tacacs_global.rb:78]: `auto_default` returns false for this property. When would it ever be nil?

File: tests/test_tacacs_global.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  def teardown
+    # teardown runs at the end of each test
+    no_tacacs_global
+    super
+  end
+
+  def no_tacacs_global
+    # Turn the feature off for a clean test.
+    config('no tacacs-server timeout 2')
```
Comment (chrisvanheuveln) [inline tests/test_tacacs_global.rb:40]: This doesn't turn off the tacacs feature. Is this a sufficient cleanup? will it clean up stale keys from a previous test?

File: tests/test_tacacs_global.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  # TESTS
+
+  def test_tacacs_global
+    id = 'default'
+
+    global = Cisco::TacacsGlobal.new(id)
+    assert_includes(Cisco::TacacsGlobal.tacacs_global, id)
+    assert_equal(Cisco::TacacsGlobal.tacacs_global[id], global)
```
Comment (chrisvanheuveln) [inline tests/test_tacacs_global.rb]: `assert_equal(`_expected, actual_`)`

**Comment by chrisvanheuveln** [comment]
@tphoney We've spoken about this before. Either: Do not use git commit --amend; or respond to each comment. We squash all commits now so you shouldn't need to use amend.

**Comment by tphoney** [comment]
@chrisvanheuveln how do you want me to proceed, are there other changes you want ?

**Comment by chrisvanheuveln** [comment]
@tphoney please respond to my earlier comments so that I know they've been addressed.

**Comment by tphoney** [comment]
/enable/tacacs/ **fixed**
/2014-2016/2016/ **fixed**
You're using both auto_default and default_value ? This should be a one-or-the-other. **auto removed**
auto_default returns false for this property. When would it ever be nil?  **auto removed**
This doesn't turn off the tacacs feature. Is this a sufficient cleanup? will it clean up stale keys from a previous test? **cleanup is sufficient for test**
assert_equal(expected, actual) **fixed**
@chrisvanheuveln see above

**Comment by chrisvanheuveln** [comment]
Please fix _all_ copyrights and _all_ assert_equals. If the tacacs cleanup is sufficient then remove the misleading comment that indicates it should do something else.

These are all minor things - please be thorough.

**Comment by tphoney** [comment]
@chrisvanheuveln All copyrights are correct in the files.
all assert_equals have been amended to.

File: tests/test_tacacs_global.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+  def test_tacacs_global
+    id = 'default'
+
+    global = Cisco::TacacsGlobal.new(id)
+    assert_includes(Cisco::TacacsGlobal.tacacs_global, id)
+    assert_equal(global, Cisco::TacacsGlobal.tacacs_global[id])
+
+    # Default Checking
+    assert_equal(global.timeout, global.default_timeout)
```
Comment (chrisvanheuveln) [inline tests/test_tacacs_global.rb]: These default checks are the ones that I was referring to...  `global.timeout` is the current state on the device; `global.default_timeout` is the expected state as defined in the yaml. You've got the order reversed, here and below.

**Comment by tphoney** [comment]
@chrisvanheuveln code amended 

> These default checks are the ones that I was referring to... global.timeout is the current state on the device; global.default_timeout is the expected state as defined in the yaml. You've got the order reversed,
> code fixed.

**Pull request merged**